copy/copy.go

```diff
@@ -655,17 +696,108 @@ func (ic *imageCopier) copyLayer(ctx context.Context, srcInfo types.BlobInfo, po
 	cachedDiffID := ic.c.blobInfoCache.UncompressedDigest(srcInfo.Digest) // May be ""
 	diffIDIsNeeded := ic.diffIDsAreNeeded && cachedDiffID == ""
 
+	// Create the initial placeholder bar which is later being replaced by
+	// another one.
+	bar := ic.c.createProgressBar(pool, srcInfo, "blob", " ", mpb.BarRemoveOnComplete())
+
 	// If we already have the blob, and we don't need to compute the diffID, then we don't need to read it from the source.
 	if !diffIDIsNeeded {
-		reused, blobInfo, err := ic.c.dest.TryReusingBlob(ctx, srcInfo, ic.c.blobInfoCache, ic.canSubstituteBlobs)
+		// tryReusingBlob is an anonymous function to avoid adding
+		// another method with a complex parameter list and semantics.
+		// It's sole purpose is to simplify the control with an early
+		// return if the image destination does not blob locking.
+		tryReusingBlob := func() (bool, types.BlobInfo, digest.Digest, error) {
+			reused, blobInfo, err := ic.c.dest.TryReusingBlob(ctx, srcInfo, ic.c.blobInfoCache, ic.canSubstituteBlobs)
+			if err != nil {
+				return false, types.BlobInfo{}, "", errors.Wrapf(err, "Error trying to reuse blob %s at destination", srcInfo.Digest)
+			}
+			if reused {
+				logrus.Debugf("Skipping blob %s (already present):", srcInfo.Digest)
+				bar.SetTotal(0, true)
+				ic.c.createProgressBar(pool, srcInfo, "blob", "skipped: already exists", mpb.BarReplaceOnComplete(bar))
+				return true, blobInfo, cachedDiffID, nil
+			}
+
+			if !ic.c.dest.SupportsBlobLocks() {
+				return false, types.BlobInfo{}, "", nil
+			}
+
+			// Check if another process might be copying the layer at the
+			// moment.  If so, wait for it's completion and try to reuse
+			// the blob again.
+			//
+			logrus.Debugf("Checking if blob %s is already being copied by another process", srcInfo.Digest)
+
+			// LockBlob() does not support try-lock semantics which would
+			// be a straight-forward mechanism to detect if another process
+			// is currently copying a specific layer.  However, with
+			// LockBlob() we are constrained by the file locks in
+			// containers/storage which cannot support try-lock semantics
+			// as fcntl(2) does not support that.
+			//
+			// The trick we are using here is to acquire the lock in a
+			// separate goroutine, which sets a boolean signal-variable as
+			// soon as the lock is acquired.  The main-routine is then
+			// sleeping for a certain amount of time.  If the lock has not
+			// been acquired until then, the progress bar will be updated
+			// with a message indicating that the blob is being copied by
+			// another process and wait until the signal-variable is set
+			// true.
+			acquiredBlobLock := false
+			go func() {
+				ic.c.dest.LockBlob(srcInfo)
```

Comment:
Looking at https://github.com/containers/storage/pull/307, this creates a new file under the storage root's new "$driver-locks" subdirectory, named using the digest of the blob, and keeps it open until the calling process exits.
How do callers that are using this storage root keep the number of digest lock files from growing without bound?  How does a long-lived caller like CRI-O avoid tying up an open file descriptor, until it exits, for each blob that it pulls down from a registry?

Reply:
The idea is to have per-digest-lock. The locks have a size of 0 as we're only syncing on the file descriptor. I agree that this is a very ugly part of the current state of the PR but I don't yet have answer to that.

Reply:
Ideally, an unlock would also delete the file but I didn't yet find a solution to avoid a race for this case.

Reply:
(for the record)
There are two issues that we address as follows:

1. Accumulating lock files in the storage: we address it by moving them into the tmp storage which won't persist reboots.
2. Too many open FDs: we'll extend the lockfiles in containers/storage to be able to close the file descriptors and perform reference counting. The last reference will close the FD. This will make long-living users such as CRI-O happier.

Reply:
Now fixed.